aten/src/ATen/native/cpu/optimizers_kernel.cpp

```diff
@@ -0,0 +1,86 @@
+#include <ATen/native/optimizers.h>
+
+#include <cmath>
+
+#include <ATen/ATen.h>
+#include <ATen/CPUApplyUtils.h>
+#include <ATen/Dispatch.h>
+#include <ATen/cpu/vec256/vec256.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename T>
+void SparseAdamStepKernelImpl(
+    T alpha,
+    T beta1,
+    T beta2,
+    T eps,
+    int64_t step,
+    TensorIterator* it) {
+  TORCH_CHECK(it->is_contiguous());
```

Comment:
this isn't what TORCH_CHECK is for, see https://github.com/pytorch/pytorch/blob/3e0da2ab8e702e20abb3cbfbd71319aa71615568/c10/util/Exception.h#L169-L191.

Also, I couldn't quickly convince myself that all of the operands are actually guaranteed to be contiguous -- can you provide an argument why (in comments?).